### DISCUSSION TITLE
Local development of a PEP561 plugin

### DISCUSSION THREAD

**Discussion by wolflu05** [inventree/InvenTree#3506] [Q&A] [answered]
Hello,

if I have such a [structure](https://github.com/inventree/inventree-brother-plugin) of plugin and want to develop locally, the easiest way is by placing this folder in a folder which is in `INVENTREE_PLUGIN_DIR`, but I didn't see the plugin appearing the the UI. Do I need something else to get that to work?

If I place a single file which contains a plugin in the `INVENTREE_PLUGIN_DIR`, it appears in the UI. What are your recommendations to develop a plugin which consists of a [PEP 561](https://peps.python.org/pep-0561/) compliant package like shown in the [documentation](https://inventree.readthedocs.io/en/latest/extend/how_to_plugin/#packaging)?

**Comment by matmair** [comment, marked as answer]
@wolflu05 this probalby will take a few posts so it probably would work better as a bug / issue ut I am going to try anyway.

There are a few things that could happen here. Please ensure all following conditions are met. This is all for development. For deploying I would strongly recomment to distribute via pypi or similar indexes.
Requirements:
- plugins are enabled
- plugin_dir points to the right directory
- `__init__.py` file in plugin_dir -> needed for module dicsovery
- `__init__.py` file in plugin subdir -> needed for module dicsovery
- inventree-server (the application, not the linux host it runs on) must be restarted once after placing the file
- specific plugin enabled in global settings\plugins or in the admin interface -> defaults to False

I started to use [editable installs](https://stackoverflow.com/a/35064498/17860466) recently. Quite handy if you work with several plugins and have the same folder structure for all your repos.

**Reply by wolflu05** [reply]
Thank you for your reply, I don't have any idea, why its not working. Everything is like you described.

```bash
(venv) vscode ➜ /workspaces/InvenTree (master ✗) $ export INVENTREE_PLUGINS_ENABLED=True
(venv) vscode ➜ /workspaces/InvenTree (master ✗) $ export INVENTREE_PLUGIN_DIR=/workspaces/InvenTree/dev/plugins
(venv) vscode ➜ /workspaces/InvenTree (master ✗) $ tree dev/plugins
dev/plugins
├── __init__.py
├── __pycache__
│   ├── __init__.cpython-310.pyc
│   └── sample.cpython-310.pyc
├── inventree-bulk-plugin
│   ├── LICENSE
│   ├── README.md
│   ├── inventree_bulk
│   │   ├── BulkGenerator
│   │   │   ├── BulkGenerator.py
│   │   │   └── dimensions.py
│   │   ├── __init__.py
│   │   ├── bulk_plugin.py
│   │   └── version.py
│   ├── setup.cfg
│   └── setup.py
└── sample.py

4 directories, 13 files
(venv) vscode ➜ /workspaces/InvenTree (master ✗) $ inv server
Waiting for database...
Database connection sucessful!
Watching for file changes with StatReloader
Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (1 silenced).
August 10, 2022 - 08:11:17
Django version 3.2.14, using settings 'InvenTree.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
```

I'm using the setup from #3474

The `sample.py` shows up in the UI, but my bulk plugin not.
![image](https://user-images.githubusercontent.com/76838159/183851675-bae90d48-3458-4714-919d-1cdfc00d2469.png)

This is the state of my current files. https://github.com/wolflu05/inventree-bulk-plugin

**Reply by wolflu05** [reply]
@matmair if I put
```txt
-e file:///workspaces/InvenTree/dev/a/inventree-bulk-plugin
```
in my plugins.txt and do `inv plugins` and `inv server` I get this error (the absolute path is correct, for testing I moved my plugin in a different dir):

<details>

```bash
(venv) vscode ➜ /workspaces/InvenTree (master ✗) $ inv server
Waiting for database...
Database connection sucessful!
Watching for file changes with StatReloader
Watching for file changes with StatReloader
Exception in thread django-main-thread:
Traceback (most recent call last):
  File "/workspaces/InvenTree/InvenTree/plugin/registry.py", line 273, in collect_plugins
    plugin._get_package_metadata()
  File "/workspaces/InvenTree/InvenTree/plugin/plugin.py", line 297, in _get_package_metadata
    meta = metadata(cls.__name__)
  File "/usr/local/lib/python3.10/importlib/metadata/__init__.py", line 974, in metadata
    return Distribution.from_name(distribution_name).metadata
  File "/usr/local/lib/python3.10/importlib/metadata/__init__.py", line 548, in from_name
    raise PackageNotFoundError(name)
importlib.metadata.PackageNotFoundError: No package metadata was found for BulkActionPlugin

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/workspaces/InvenTree/InvenTree/plugin/helpers.py", line 68, in handle_error
    package_name = pathlib.Path(package_path).relative_to(install_path).parts[0]
  File "/usr/local/lib/python3.10/pathlib.py", line 816, in relative_to
    raise ValueError("{!r} is not in the subpath of {!r}"
ValueError: '/usr/local/lib/python3.10/importlib/metadata/__init__.py' is not in the subpath of '/workspaces/InvenTree/venv/lib/python3.10/site-packages' OR one path is relative and the other is absolute.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/threading.py", line 1009, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.10/threading.py", line 946, in run
    self._target(*self._args, **self._kwargs)
  File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/utils/autoreload.py", line 64, in wrapper
    fn(*args, **kwargs)
  File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/core/management/commands/runserver.py", line 110, in inner_run
    autoreload.raise_last_exception()
  File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/utils/autoreload.py", line 87, in raise_last_exception
    raise _exception[1]
  File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 375, in execute
    autoreload.check_errors(django.setup)()
  File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/utils/autoreload.py", line 64, in wrapper
    fn(*args, **kwargs)
  File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/apps/registry.py", line 122, in populate
    app_config.ready()
  File "/workspaces/InvenTree/InvenTree/plugin/apps.py", line 46, in ready
    registry.collect_plugins()
  File "/workspaces/InvenTree/InvenTree/plugin/registry.py", line 276, in collect_plugins
    handle_error(error, do_raise=False, log_name='discovery')
  File "/workspaces/InvenTree/InvenTree/plugin/helpers.py", line 71, in handle_error
    path_obj = pathlib.Path(package_path).relative_to(settings.BASE_DIR)
  File "/usr/local/lib/python3.10/pathlib.py", line 816, in relative_to
    raise ValueError("{!r} is not in the subpath of {!r}"
ValueError: '/usr/local/lib/python3.10/importlib/metadata/__init__.py' is not in the subpath of '/workspaces/InvenTree/InvenTree' OR one path is relative and the other is absolute.
```

</details>

**Reply by matmair** [reply]
> Thank you for your reply, I don't have any idea, why its not working. Everything is like you described.
> 
> ```bash
> (venv) vscode ➜ /workspaces/InvenTree (master ✗) $ export INVENTREE_PLUGINS_ENABLED=True
> (venv) vscode ➜ /workspaces/InvenTree (master ✗) $ export INVENTREE_PLUGIN_DIR=/workspaces/InvenTree/dev/plugins
> (venv) vscode ➜ /workspaces/InvenTree (master ✗) $ tree dev/plugins
> dev/plugins
> ├── __init__.py
> ├── __pycache__
> │   ├── __init__.cpython-310.pyc
> │   └── sample.cpython-310.pyc
> ├── inventree-bulk-plugin
> │   ├── LICENSE
> │   ├── README.md
> │   ├── inventree_bulk
> │   │   ├── BulkGenerator
> │   │   │   ├── BulkGenerator.py
> │   │   │   └── dimensions.py
> │   │   ├── __init__.py
> │   │   ├── bulk_plugin.py
> │   │   └── version.py
> │   ├── setup.cfg
> │   └── setup.py
> └── sample.py
> 
> 4 directories, 13 files
> (venv) vscode ➜ /workspaces/InvenTree (master ✗) $ inv server
> Waiting for database...
> Database connection sucessful!
> Watching for file changes with StatReloader
> Watching for file changes with StatReloader
> Performing system checks...
> 
> System check identified no issues (1 silenced).
> August 10, 2022 - 08:11:17
> Django version 3.2.14, using settings 'InvenTree.settings'
> Starting development server at http://127.0.0.1:8000/
> Quit the server with CONTROL-C.
> ```
> 
> I'm using the setup from #3474
> 
> The `sample.py` shows up in the UI, but my bulk plugin not.
> ![image](https://user-images.githubusercontent.com/76838159/183851675-bae90d48-3458-4714-919d-1cdfc00d2469.png)
> 
> This is the state of my current files. https://github.com/wolflu05/inventree-bulk-plugin

@wolflu05 you are missing a __init__.py in inventree-bulk-plugin

**Reply by matmair** [reply]
> @matmair if I put
> ```txt
> -e file:///workspaces/InvenTree/dev/a/inventree-bulk-plugin
> ```
> in my plugins.txt and do `inv plugins` and `inv server` I get this error (the absolute path is correct, for testing I moved my plugin in a different dir):
> 
> <details>
> 
> ```bash
> (venv) vscode ➜ /workspaces/InvenTree (master ✗) $ inv server
> Waiting for database...
> Database connection sucessful!
> Watching for file changes with StatReloader
> Watching for file changes with StatReloader
> Exception in thread django-main-thread:
> Traceback (most recent call last):
>   File "/workspaces/InvenTree/InvenTree/plugin/registry.py", line 273, in collect_plugins
>     plugin._get_package_metadata()
>   File "/workspaces/InvenTree/InvenTree/plugin/plugin.py", line 297, in _get_package_metadata
>     meta = metadata(cls.__name__)
>   File "/usr/local/lib/python3.10/importlib/metadata/__init__.py", line 974, in metadata
>     return Distribution.from_name(distribution_name).metadata
>   File "/usr/local/lib/python3.10/importlib/metadata/__init__.py", line 548, in from_name
>     raise PackageNotFoundError(name)
> importlib.metadata.PackageNotFoundError: No package metadata was found for BulkActionPlugin
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/workspaces/InvenTree/InvenTree/plugin/helpers.py", line 68, in handle_error
>     package_name = pathlib.Path(package_path).relative_to(install_path).parts[0]
>   File "/usr/local/lib/python3.10/pathlib.py", line 816, in relative_to
>     raise ValueError("{!r} is not in the subpath of {!r}"
> ValueError: '/usr/local/lib/python3.10/importlib/metadata/__init__.py' is not in the subpath of '/workspaces/InvenTree/venv/lib/python3.10/site-packages' OR one path is relative and the other is absolute.
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.10/threading.py", line 1009, in _bootstrap_inner
>     self.run()
>   File "/usr/local/lib/python3.10/threading.py", line 946, in run
>     self._target(*self._args, **self._kwargs)
>   File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/utils/autoreload.py", line 64, in wrapper
>     fn(*args, **kwargs)
>   File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/core/management/commands/runserver.py", line 110, in inner_run
>     autoreload.raise_last_exception()
>   File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/utils/autoreload.py", line 87, in raise_last_exception
>     raise _exception[1]
>   File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 375, in execute
>     autoreload.check_errors(django.setup)()
>   File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/utils/autoreload.py", line 64, in wrapper
>     fn(*args, **kwargs)
>   File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/__init__.py", line 24, in setup
>     apps.populate(settings.INSTALLED_APPS)
>   File "/workspaces/InvenTree/venv/lib/python3.10/site-packages/django/apps/registry.py", line 122, in populate
>     app_config.ready()
>   File "/workspaces/InvenTree/InvenTree/plugin/apps.py", line 46, in ready
>     registry.collect_plugins()
>   File "/workspaces/InvenTree/InvenTree/plugin/registry.py", line 276, in collect_plugins
>     handle_error(error, do_raise=False, log_name='discovery')
>   File "/workspaces/InvenTree/InvenTree/plugin/helpers.py", line 71, in handle_error
>     path_obj = pathlib.Path(package_path).relative_to(settings.BASE_DIR)
>   File "/usr/local/lib/python3.10/pathlib.py", line 816, in relative_to
>     raise ValueError("{!r} is not in the subpath of {!r}"
> ValueError: '/usr/local/lib/python3.10/importlib/metadata/__init__.py' is not in the subpath of '/workspaces/InvenTree/InvenTree' OR one path is relative and the other is absolute.
> ```
> 
> </details>

Editable installs currently do not work with the plugin file - just run pip install -e in a shell in the same environment that your instance is using.

**Reply by wolflu05** [reply]
> Editable installs currently do not work with the plugin file - just run pip install -e in a shell in the same environment that your instance is using.

@matmair Same error occurs

**Reply by wolflu05** [reply]
My plugin has the same structure like this official one https://github.com/inventree/inventree-brother-plugin

**Reply by wolflu05** [reply]
@matmair I'd be happy if you can help me with this. By placing another init file the discovery works, but the module resolution not. I'd really appreciate it, if you can try that also out on your side.
![image](https://user-images.githubusercontent.com/76838159/183985533-acad8ace-be10-4dd5-967a-204a80501a84.png)

**Reply by matmair** [reply]
> My plugin has the same structure like this official one https://github.com/inventree/inventree-brother-plugin

This is not optimised for directory dev but editable installs. I do not really recommend folder development as it does not behave the same as deployment.

I will take a look at your plugin once I find time to spin up a full on prem enviroment for testing again.

A general note about python discovery: You need a `__init__.py` in every folder that should be dicsoverable. Python does not look deeper if that file is not there. [Here](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#automatic-discovery) is extensive context to that.

**Reply by wolflu05** [reply]
@matmair Ok, so I should focus on the editable install. I run it in the devcontainer through vscode, not on a on prem install for development.

**Reply by matmair** [reply]
@wolflu05 I submitted a PR that fixes detection. Editable is the way to go though. Please mark the thread as answer if everything runs now.

**Reply by wolflu05** [reply]
@matmair Thank you very much, now it's working.
But one thing is not clear to me, what is the purpose of this signature field? And how can I get my plugin a signature? 
![image](https://user-images.githubusercontent.com/76838159/184122182-5f8b0fe3-3327-4242-a9b3-fb8ee40c5466.png)

**Reply by matmair** [reply]
it is for gpg signatures - if you signed your commits with gpg it will show up here. I am actually thinking about disabeling signature information by default.

**Reply by wolflu05** [reply]
@matmair This does not really work, as you can see, my last commits in dev are signed and I still have this red box with the no signature warning.

**Reply by matmair** [reply]
This only works if you:
- a) included git metadata
- b) have gpg installed in our env

I am going to disable this by default becaues it only creates confusion and only a select few users care about this at all.

**Reply by wolflu05** [reply]
Ok. Thank you for clarification.
